nixos/modules/programs/wayland/niri.nix

```diff
@@ -0,0 +1,73 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.programs.niri;
+in
+{
+  options.programs.niri = {
+    enable = lib.mkEnableOption "Niri, a scrollable-tiling Wayland compositor";
+
+    package = lib.mkPackageOption pkgs "niri" { };
+
+    # Recommended by upstream
+    # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+    portalPackage = lib.mkPackageOption pkgs "xdg-desktop-portal-gnome" { };
+  };
+
+  config = lib.mkIf cfg.enable (
+    lib.mkMerge [
+      {
+        environment = {
+          systemPackages = [ cfg.package ];
+        };
+
+        services = {
+          displayManager.sessionPackages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#portals
+          gnome.gnome-keyring.enable = lib.mkDefault true;
+        };
+
+        systemd = {
+          packages = [ cfg.package ];
+
+          # Recommended by upstream
+          # https://github.com/YaLTeR/niri/wiki/Important-Software#authentication-agent
+          user.services.niri-polkit-agent = lib.mkIf config.security.polkit.enable {
```

Comment:
> is this really suitable for nixpkgs?

Considering this is actually a 1:1 copy of `${pkgs.polkit-kde-agent}/lib/systemd/user/plasma-polkit-agent.service`, I'd assume so

> Are you confident that this works smoothly?

I can use `pkexec` and `run0` just fine. This should work the same as it does in the Plasma module

Reply:
Wouldn't it be sufficient then to add that package to `systemd.packages` and just add the dependency from niri to it here?

Reply:
Let me rephrase: it's *almost* a 1:1 copy. We still need to replace the Plasma units with Niri's

Reply:
> > is this really suitable for nixpkgs?
> 
> Considering this is actually a 1:1 copy of `${pkgs.polkit-kde-agent}/lib/systemd/user/plasma-polkit-agent.service`, I'd assume so

yeah, but i mean since no other "simple" wayland compositor comes with a polkit agent, why does niri get one? what is the polkit story for those on NixOS, anyway? what does one need to do to get a smoothly working polkit setup on e.g. Sway or Hyprland? is there actually a good reason to include this for niri in particular? the plasma polkit agent makes a lot of sense because *that polkit agent is made for that DE* and a DE is supposed to always include all these parts out of the box. niri doesn't necessarily fit either of those descriptions, and although niri does recommend using the KDE polkit agent, there's actually no good objective reason why you must use that one, it's not inherently linked to niri, unlike `xdg-desktop-portal-gnome` which *is* actually required for certain functionality (screencast, some output management in gnome control center) and directly interfaces with niri in a way no other xdg-desktop-portal could. that one is actually a dependency (albeit not a *hard* dependency; you can drop it and just Not Have Screencast, but there's no alternative), whereas the KDE polkit agent is just a "recommended software to go along with" like fuzzel, alacritty, swaybg, mako, and waybar; all of which can also reasonably be replaced with equivalents and you'll have the same functionality, actually

> > Are you confident that this works smoothly?
> 
> I can use `pkexec` and `run0` just fine. This should work the same as it does in the Plasma module

yeah, i'm sure polkit itself works fine, but that's not what i meant, though. maybe i wasn't clear, but i'm more concerned as to how it interacts with session management and service lifecycles. to be clear, not all of that is the fault of a polkit agent, but i'm just... tired... of dealing with systemd issues... and i know that's vague as hell, but i genuinely don't know the root cause, and this just has Bad Vibes because my polkit agent is systemd-janky and this looks a lot like it. i really don't want to merge something that will cause subtle, hard-to-debug issues, for all the existing niri users.

when i ask if this works smoothly, i mean can you run a full niri session with whatever auxiliary services (swaybg/swww, swayidle/hypridle, waybar/eww, swaync/mako/dunst, and such) and They Just Start Fine Without Issues? because right now i think my config has an all-time low number of issues, but from a quick search [i still seem to have one `swayidle`-related hack left](https://github.com/sodiboo/system/blob/2978f4d79c51a5bd7e38a9cd75e3ec9046aa7e75/lock.mod.nix#L123-L126). that question seems fairly disconnected from polkit, but i promise you, this is all a interconected with the most annoying relations, i've been tearing my hair out for months over this, and i just don't want to make it Worse.

Reply:
> yeah, but i mean since no other "simple" wayland compositor comes with a polkit agent, why does niri get one? what is the polkit story for those on NixOS, anyway? what does one need to do to get a smoothly working polkit setup on e.g. Sway or Hyprland? is there actually a good reason to include this for niri in particular?

I can't really speak for other compositors, I just figured it would be nice here. It's recommended by upstream, already included in niri-flake for example, and it seems to be a hard thing to implement given all of this discussion. I can understand wanting to leave this up to the user, though -- I would just want some discussion about it rather than only "other modules don't"

> and although niri does recommend using the KDE polkit agent, there's actually no good objective reason why you must use that one

You can override the `ExecStart` here to change it pretty easily

> yeah, i'm sure polkit itself works fine, but that's not what i meant, though.

This is talking about the Polkit *agent* working fine, not just polkit itself

>  i really don't want to merge something that will cause subtle, hard-to-debug issues, for all the existing niri users

I haven't come across these in testing. We should probably get other people to check of course, but if the above commands work I'm not sure how this agent could fail elsewhere

> when i ask if this works smoothly, i mean can you run a full niri session with whatever auxiliary services (swaybg/swww, swayidle/hypridle, waybar/eww, swaync/mako/dunst, and such) and They Just Start Fine Without Issues?

If the service files for those services are written correctly, I'd think so
But I'm not sure what this has to do with us getting the Polkit agent's service right?

Reply:
Let's just add an enable option so people can disable it, and move on?